lib/query-interface.js

```diff
@@ -495,18 +495,21 @@ QueryInterface.prototype.insert = function(instance, tableName, values, options)
   });
 };
 
-QueryInterface.prototype.upsert = function(tableName, values, updateValues, model, options) {
+QueryInterface.prototype.upsert = function(tableName, valuesByField, updateValues, model, options) {
   var wheres = []
     , where
     , indexFields
     , indexes = []
-    , attributes = Object.keys(values);
+    , attributes = Object.keys(valuesByField);
 
   where = {};
-  for (var i = 0; i < model.primaryKeyAttributes.length; i++) {
+
+  for (var i = 0; i < model.primaryKeyAttributes.length; i++){
     var key = model.primaryKeyAttributes[i];
-    if(key in values){
-      where[key] = values[key];
+    var attribute = model.rawAttributes[key];
+    var fieldName = attribute.field || attribute.fieldName;
+    if(fieldName in valuesByField){
+      where[fieldName] = valuesByField[fieldName];
     }
   }
```

Comment:
Updated ... It's your code :) Reduce seems like a strange choice to me in this case.